HACKING.rst

```diff
@@ -425,11 +426,180 @@ This can be done incrementally, one function at a time:
 * refactor its implementation from ``cloudinit.net`` into the
   ``Networking`` hierarchy (e.g. if it has an if/else on BSD, this is
   the time to put the implementations in their respective subclasses)
+
+  * if part of the method contains distro-independent logic, then you
+    may need to create new methods to capture this distro-specific
+    logic; we don't want to replicate common logic in different
+    ``Networking`` subclasses
+  * if after the refactor, the method on the root ``Networking`` class
+    no longer has any implementation, it should be converted to an
+    `abstractmethod`_
+
 * ensure that the new implementation has unit tests (either by moving
   existing tests, or by writing new ones)
+* ensure that the new implementation has a docstring
 * finally, remove it (and any other now-unused functions) from
   cloudinit.net (to avoid having two parallel implementations)
 
+``cloudinit.net`` Functions/Classes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The functions/classes that need refactoring break down into some broad
+categories:
+
+* helpers for accessing ``/sys`` (that should not be on the top-level
+  ``Networking`` class as they are Linux-specific):
+
+  * ``get_sys_class_path``
+  * ``sys_dev_path``
+  * ``read_sys_net``
+  * ``read_sys_net_safe``
+  * ``read_sys_net_int``
+
+* those that directly access ``/sys`` (via helpers) and should (IMO) be
+  included in the API of the ``Networking`` class:
+
+  * ``generate_fallback_config``
+
+    * the ``config_driver`` parameter is used and passed as a boolean,
+      so we can change the default value to ``False`` (instead of
+      ``None``)
+
+  * ``get_ib_interface_hwaddr``
+  * ``get_interface_mac``
+  * ``interface_has_own_mac``
+  * ``is_bond``
+  * ``is_bridge``
+  * ``is_connected``
+  * ``is_physical``
+  * ``is_present``
+  * ``is_renamed``
+  * ``is_up``
+  * ``is_vlan``
+  * ``is_wireless``
+  * ``wait_for_physdevs``
+
+* those that directly access ``/sys`` (via helpers) but may be
+  Linux-specific concepts or names:
+
+  * ``get_master``
+  * ``device_devid``
+  * ``device_driver``
+
+* those that directly use ``ip``:
+
+  * ``_get_current_rename_info``
+
+    * this has non-distro-specific logic so should potentially be
+      refactored to use helpers on ``self`` instead of ``ip`` directly
+      (rather than being wholesale reimplemented in each of
+      ``BSDNetworking`` or ``LinuxNetworking``)
```

Comment:
The nic renaming bit might be Linux specific; not sure if BSD has similar issues with Linux where the eth* kernel device namespace has issues with being renamed by user-space.

Reply:
Right, that's a good point. @igalic @goneri can you let us know what you think about network device renaming on *BSD?

Reply:
the main reason we do it is because it helps us reuse the vendor's cloud-config, which usually uses `eth0`, even though they of all people should know what the devices will be named

Reply:
I'm convinced no BSD user wants to set the interface renamed on the fly, especially to something super Linux-ish like `eth0`. Today, the feature only works on FreeBSD.

Reply:
Excluding user-generated networking configs; the primary reason for renaming nics is to bind a provided network config to a specific interface (identified by MAC).

Some virtualization platforms do not guarantee that the device name of the NIC is stable across reboots; that means eth0 with macA may be eth1 on the next boot, or in the case of Linux systemd, the "persistent name" changes as the PCI bus slot the NIC was attached to has changed (user added a virtual interface to their instance).

For containers, the Linux specific rules generated to ensure the name to mac mapping remains across reboots (udev /etc/udev/rules.d/70-persistent-net) does not operate inside containers; yet cloud-init still needs to ensure that the provided network config is on the correct interface.

@goneri any comments w.r.t to multi-nic VMs on public clouds and any experience with nic name changing due to device enumeration randomness?  Originally this was seen on Amazon Xen-based instances.

Reply:
> I'm convinced no BSD user wants to see the interface renamed on the fly

I'm not sure many Linux users particularly want this either but, as Ryan lays out, the reality of cloud networking (on Linux, at least) sometimes requires it.

(And it's worth bearing in mind that if someone considers themselves a cloud-init user, or a $cloud user, then they may want consistency across distros/OSes more than they want the most appropriate behaviour for a particular OS or distro. This isn't a trump card argument, it's just another angle to consider.)